lib/winapi/sync.c

```diff
@@ -38,6 +40,377 @@ VOID LeaveCriticalSection (LPCRITICAL_SECTION lpCriticalSection)
+                break;
+            case INITONCE_ASYNC_IN_PROGRESS:
+                if (!(dwFlags & INIT_ONCE_ASYNC)) {
+                    SetLastError(ERROR_INVALID_PARAMETER);
+                    *fPending = TRUE;
+                    return FALSE;
+                }
+                *fPending = TRUE;
+                return TRUE;
+            case INITONCE_DONE:
+                if (lpContext) {
+                    *lpContext = (PVOID)((ULONG_PTR)lpInitOnce->Ptr & ~INITONCE_MASK);
+                }
+                *fPending = FALSE;
+                return TRUE;
+        }
+    }
+}
+
+BOOL InitOnceExecuteOnce (PINIT_ONCE InitOnce, PINIT_ONCE_FN InitFn, PVOID Parameter, LPVOID *Context)
+{
+    if ((((ULONG_PTR)Context) & INITONCE_MASK) != INITONCE_UNINITIALIZED) {
+        SetLastError(ERROR_INVALID_PARAMETER);
+        return FALSE;
+    }
+
+    BOOL fPending;
+    BOOL ret;
+    ret = InitOnceBeginInitialize(InitOnce, 0, &fPending, Context);
+    if (!fPending) {
+        return ret;
+    }
+
+    if (InitFn(InitOnce, Parameter, Context)) {
+        return InitOnceComplete(InitOnce, 0, Context ? *Context : NULL);
+    }
+
+    InitOnceComplete(InitOnce, INIT_ONCE_INIT_FAILED, NULL);
+    return FALSE;
+}
+
+BOOL InitOnceComplete (LPINIT_ONCE lpInitOnce, DWORD dwFlags, LPVOID lpContext)
+{
+    if (((ULONG_PTR)lpContext & INITONCE_MASK) != INITONCE_UNINITIALIZED) {
+        SetLastError(ERROR_INVALID_PARAMETER);
+        return FALSE;
+    }
+
+    if (dwFlags & INIT_ONCE_INIT_FAILED) {
+        if (lpContext || (dwFlags & INIT_ONCE_ASYNC)) {
+            SetLastError(ERROR_INVALID_PARAMETER);
+            return FALSE;
+        }
+    } else {
+        lpContext = (PVOID)((ULONG_PTR)lpContext | INITONCE_DONE);
+    }
+
+    while (true) {
+        PVOID cur_val = lpInitOnce->Ptr;
+
+        switch ((ULONG_PTR)lpInitOnce->Ptr & INITONCE_MASK) {
+            case INITONCE_IN_PROGRESS:
+                if (__sync_val_compare_and_swap(&lpInitOnce->Ptr, cur_val, lpContext) != cur_val) {
+                    break; // failed to swap the value, try again
+                }
+
+                // init done, waiters continue because they were spinning in InitOnceBeginInitialize
+                return TRUE;
+            case INITONCE_ASYNC_IN_PROGRESS:
+                if(!(dwFlags & INIT_ONCE_ASYNC)) {
+                    SetLastError(ERROR_INVALID_PARAMETER);
+                    return FALSE;
+                }
+
+                if (__sync_val_compare_and_swap(&lpInitOnce->Ptr, cur_val, lpContext) == cur_val) {
+                    // async init done
+                    return TRUE;
+                }
+                break;
+            default:
+                SetLastError(ERROR_GEN_FAILURE);
+                return FALSE;
+        }
+    }
+}
+
+static BOOL WINAPI condvar_init (PINIT_ONCE InitOnce, PVOID Parameter, PVOID *Context)
+{
+    PCONDITION_VARIABLE ConditionVariable = Parameter;
+    NTSTATUS status;
```

Comment:
nit: Before creating new events, you could probably check if this thing is really clean at this point:

- `ConditionVariable->initOnce.Ptr == 0`
- `ConditionVariable->waitCount == 0`
- `ConditionVariable->eventHandles[0] == INVALID_HANDLE_VALUE`
- `ConditionVariable->eventHandles[1] == INVALID_HANDLE_VALUE`

You'd just have to reintroduce the initialization of the events to `INVALID_HANDLE_VALUE` in `InitializeConditionVariable`.
These assumptions should always be true then, because they come from `CONDITION_VARIABLE_INIT` or `InitializeConditionVariable`.

However, I'll leave up to you wether you consider this a useful addition that's worth adding.

Reply:
Ah, I forgot about this.
We can't really check for `ConditionVariable->initOnce.Ptr == 0` here, because the InitOnce API functions will change the value of this variable before calling this function, so it's never zero here.
I did add checks for the other conditions though.